[FIX] Clarify data types and requirement levels for all JSON files

sappelhoff opened the pull request:
closes #350 
closes #221 
closes #533 
fixes part of https://github.com/bids-standard/bids-validator/issues/1065

this is a draft to get comments on whether this would be something we want for the entire spec.

Also some specific questions:

- Does somebody have a better link to describe JSON strings than https://www.w3schools.com/js/js_json_syntax.asp ?
- What do you think of the "data type" column entry for `Levels` --> `[object][object] with string: string key-value pairs` is that clear enough? --> in JSON all keys MUST be strings ... but I wanted to make the point that the values in the `Levels` object MUST be strings as well.

# TO DO

- [x] render pdf in landscape so that tables are easier to read
- [x] decrease the extremely large margins in the PDF build (currently 1.5 inch), so that tables are displayed more easily --> perhaps go to 0.75 inch
    - [x] ~center~ align the "fancyheader" and "fancyfooter" (see `header.tex`) to make the PDF look nice with smaller marging
- [x] make sure that no tables break in the pdf build (see e.g., https://github.com/bids-standard/bids-specification/pull/605/files#r488467181) --> **NOTE**: Some single rows *do* break, e.g., `AnatomicalLandmarkCoordinateSystemDescription` ... simply because the "key" is too long and will overlap with the req. level column. that happens only a few times and is unavoidable at this point.
- [x] double check that all tables render in the html build

# PDF status

~I don't know what's happening ... in a8cb2499912db8f5ebd294bcdf2120545ba02d47 everything is nice ... one commit later the tables are broken again: 357bfd3f7f258a19e77b96830ae9c7599f633992 (--> checkout the circle ci build artifact in each commit)~

--> fixed. This was a funny mix of python's `os.walk` not being deterministic, a broken MD table, and a python-table-rendering script that did not warn upon error :)

tsalo reviewed (approved):
I think the level of detail is quite reasonable.

tsalo commented:
@sappelhoff do you want to convert the PR to a draft to make sure no one accidentally merges it once reviews are in?

effigies commented:
This looks reasonable.

1. I don't know of a better reference for strings.
2. We could say `[object][] of [strings][string]`, as examples should make clear what that means, but being explicit about `key/value` pairs is also fine.

sappelhoff commented on src/05-derivatives/03-imaging.md:
```diff
@@ -131,13 +131,13 @@ and the `Atlas` metadata SHOULD be defined.
 
-| **Key name** | **Description**                                                                                                                                |
-| ------------ | ---------------------------------------------------------------------------------------------------------------------------------------------- |
-| RawSources   | Same as defined in [Introduction][intro], but elevated from OPTIONAL to REQUIRED                                                               |
```
@effigies could this be an error here? RawSources is defined in `02-common-data-types.md`, but not in `01-introduction.md`.

effigies replied:
Yes, looks like it broke during a reorganization.

sappelhoff commented on pdf_build_src/header_setup.tex:
```diff
@@ -25,3 +25,5 @@
     basewidth=0.5em
 }
+
+\usepackage[a4paper, margin=1.5in]{geometry}
```
not sure how, but this change may break several tables :man_shrugging: 

- see p.63 and down in [previous pdf](https://circle-production-customer-artifacts.s3.amazonaws.com/picard/5bac04e913babe000192f2a5/5f5a9975636a043aa45b3639-0-build/artifacts/bids-spec.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20200915T075534Z&X-Amz-SignedHeaders=host&X-Amz-Expires=60&X-Amz-Credential=AKIAJR3Q6CR467H7Z55A%2F20200915%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=094061e0fc6a01337b75f60bf4a8229fa863e3f9a30ba1455be48a8a1cbc5e32)
- compared to p49 and scroll down in [this pdf](https://circle-production-customer-artifacts.s3.amazonaws.com/picard/forks/5bac04e913babe000192f2a5/151142273/5f60717f651d6e58c50eae60-0-build/artifacts/bids-spec.pdf?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Date=20200915T075059Z&X-Amz-SignedHeaders=host&X-Amz-Expires=60&X-Amz-Credential=AKIAJR3Q6CR467H7Z55A%2F20200915%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Signature=ca59a73cce238fa9f2e220d356e9a4a82ef22a974f62aebc94606da29206e2b4)

@sebastientourbier do you have a clue about what's going on here?

I'd like the tables to be nice again, and potentially also decrease the marging of the whole document by a bit

sappelhoff replied:
fixed it, was a tricky mix of bugs, described it in the OP.

effigies commented on src/04-modality-specific-files/03-electroencephalography.md:
```diff
@@ -91,56 +91,56 @@ be specified.
-| SamplingFrequency  | REQUIRED. Sampling frequency (in Hz) of all the data in the recording, regardless of their type (e.g., 2400).                                                                                                                                                                                                                                                      |
-| PowerLineFrequency | REQUIRED. Frequency (in Hz) of the power grid at the geographical location of the EEG instrument (i.e., 50 or 60).                                                                                                                                                                                                                                                 |
-| SoftwareFilters    | REQUIRED. [Object][] of temporal software filters applied, or `"n/a"` if the data is not available. Each key:value pair in the JSON object is a name of the filter and an object in which its parameters are defined as key:value pairs. E.g., `{"Anti-aliasing filter": {"half-amplitude cutoff (Hz)": 500, "Roll-off": "6dB/Octave"}}`.                          |
+| **Key name**       | **Requirement level** | **Data type**                        | **Description**                                                                                                                                                                                                                                                                                                           ...
+| ------------------ | --------------------- | ------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| EEGReference       | REQUIRED              | [string][]                           | General description of the reference scheme used and (when applicable) of location of the reference electrode in the raw recordings (e.g., "left mastoid", "Cz", "CMS"). If different channels have a different reference, this field should have a general description and the channel specific reference should be defi ...
+| SamplingFrequency  | REQUIRED              | [number][]                           | Sampling frequency (in Hz) of all the data in the recording, regardless of their type (e.g., 2400).                                                                                                                                                                                                                       ...
+| PowerLineFrequency | REQUIRED              | [integer][] or `"n/a"`               | Frequency (in Hz) of the power grid at the geographical location of the EEG instrument (i.e., 50 or 60).                                                                                                                                                                                                                  ...
+| SoftwareFilters    | REQUIRED              | [object][] of [objects][] or `"n/a"` | [Object][] of temporal software filters applied, or `"n/a"` if the data is not available. Each key:value pair in the JSON object is a name of the filter and an object in which its parameters are defined as key:value pairs. E.g., `{"Anti-aliasing filter": {"half-amplitude cutoff (Hz)": 500, "Roll-off": "6dB/Octav ...
```
Adding backticks to `channels.tsv` and two spaces/dashes to all other lines.

```suggestion
| **Key name**       | **Requirement level** | **Data type**                        | **Description**                                                                                                                                                                                                                                                                                                                                          |
| ------------------ | --------------------- | ------------------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| EEGReference       | REQUIRED              | [string][]                           | General description of the reference scheme used and (when applicable) of location of the reference electrode in the raw recordings (e.g., "left mastoid", "Cz", "CMS"). If different channels have a different reference, this field should have a general description and the channel specific reference should be defined in the `channels.tsv` file. |
| SamplingFrequency  | REQUIRED              | [number][]                           | Sampling frequency (in Hz) of all the data in the recording, regardless of their type (e.g., 2400).                                                                                                                                                                                                                                                      |
| PowerLineFrequency | REQUIRED              | [integer][] or `"n/a"`               | Frequency (in Hz) of the power grid at the geographical location of the EEG instrument (i.e., 50 or 60).                                                                                                                                                                                                                                                 |
| SoftwareFilters    | REQUIRED              | [object][] of [objects][] or `"n/a"` | [Object][] of temporal software filters applied, or `"n/a"` if the data is not available. Each key:value pair in the JSON object is a name of the filter and an object in which its parameters are defined as key:value pairs. E.g., `{"Anti-aliasing filter": {"half-amplitude cutoff (Hz)": 500, "Roll-off": "6dB/Octave"}}`.                          |
```

VisLab replied:
Would you consider having PowerLineFrequency a number rather than an integer. Line noise isn't always exactly at its nominal value. 

Is there any planning for having more specific validation criteria such as "positive", "non-negative", "<X", etc.  Just curious?

sappelhoff replied:
> Adding backticks to channels.tsv and two spaces/dashes to all other lines.

thanks

> Would you consider having PowerLineFrequency a number rather than an integer. Line noise isn't always exactly at its nominal value.

True, will adjust.

> Is there any planning for having more specific validation criteria such as "positive", "non-negative", "<X", etc. Just curious?

that'd be pretty cool actually. We are using the JSON schema in the validator, I haven't looked into whether or not such validation is possible

sappelhoff replied:
indeed, validating that "PowerLineFrequency" is positive we can use `"exclusiveMinimum": 0` in the JSON schema, see the [docs for that](https://json-schema.org/understanding-json-schema/reference/numeric.html#range). I am making a PR to bids-validator to do just that.

sappelhoff commented:
I finished all tables throughout the spec. for the MRI tables there were a couple of fields where I couldn't infer the data type. Help would be appreciated - perhaps @yarikoptic can have a look?

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| Manufacturer                  | RECOMMENDED           | [string][]    | Manufacturer of the equipment that produced the composite instances. Corresponds to DICOM Tag 0008, 0070 `Manufacturer`                                                                                                                                                   |
+| ManufacturersModelName        | RECOMMENDED           | [string][]    | Manufacturer's model name of the equipment that produced the composite instances. Corresponds to DICOM Tag 0008, 1090 `Manufacturers Model Name`                                                                                                                          |
+| DeviceSerialNumber            | RECOMMENDED           | [string][]    | The serial number of the equipment that produced the composite instances. Corresponds to DICOM Tag 0018, 1000 `DeviceSerialNumber`. A pseudonym can also be used to prevent the equipment from being identifiable, so long as each pseudonym is unique within the dataset |
+| StationName                   | RECOMMENDED           |               | Institution defined name of the machine that produced the composite instances. Corresponds to DICOM Tag 0008, 1010 `Station Name`                                                                                                                                         |
```
what datatype is this?

effigies replied:
Every instance I can quickly find in OpenNeuro is a string, which matches my intuition.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| StationName                   | RECOMMENDED           |               | Institution defined name of the machine that produced the composite instances. Corresponds to DICOM Tag 0008, 1010 `Station Name`                                                                                                                                         |
+| SoftwareVersions              | RECOMMENDED           | [string][]    | Manufacturer's designation of software version of the equipment that produced the composite instances. Corresponds to DICOM Tag 0018, 1020 `Software Versions`                                                                                                            |
+| HardcopyDeviceSoftwareVersion | [DEPRECATED][]        | [string][]    | Manufacturer's designation of the software of the device that created this Hardcopy Image (the printer). Corresponds to DICOM Tag 0018, 101A `Hardcopy Device Software Version`                                                                                           |
+| MagneticFieldStrength         | RECOMMENDED           |               | Nominal field strength of MR magnet in Tesla. Corresponds to DICOM Tag 0018,0087 `Magnetic Field Strength`                                                                                                                                                                |
```
what datatype is this?

effigies replied:
Number. In OpenNeuro, it's mostly numbers, though there are a couple strings like `"3T"` and `" "`. I would call those invalid, though.

sappelhoff replied:
yarik seems to agree: https://github.com/bids-standard/bids-validator/issues/851

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| SoftwareVersions              | RECOMMENDED           | [string][]    | Manufacturer's designation of software version of the equipment that produced the composite instances. Corresponds to DICOM Tag 0018, 1020 `Software Versions`                                                                                                            |
+| HardcopyDeviceSoftwareVersion | [DEPRECATED][]        | [string][]    | Manufacturer's designation of the software of the device that created this Hardcopy Image (the printer). Corresponds to DICOM Tag 0018, 101A `Hardcopy Device Software Version`                                                                                           |
+| MagneticFieldStrength         | RECOMMENDED           |               | Nominal field strength of MR magnet in Tesla. Corresponds to DICOM Tag 0018,0087 `Magnetic Field Strength`                                                                                                                                                                |
+| ReceiveCoilName               | RECOMMENDED           |               | Information describing the receiver coil. Corresponds to DICOM Tag 0018, 1250 `Receive Coil Name`, although not all vendors populate that DICOM Tag, in which case this field can be derived from an appropriate private DICOM field                                      |
```
what datatype is this?

effigies replied:
All strings in OpenNeuro. I concur.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| HardcopyDeviceSoftwareVersion | [DEPRECATED][]        | [string][]    | Manufacturer's designation of the software of the device that created this Hardcopy Image (the printer). Corresponds to DICOM Tag 0018, 101A `Hardcopy Device Software Version`                                                                                           |
+| MagneticFieldStrength         | RECOMMENDED           |               | Nominal field strength of MR magnet in Tesla. Corresponds to DICOM Tag 0018,0087 `Magnetic Field Strength`                                                                                                                                                                |
+| ReceiveCoilName               | RECOMMENDED           |               | Information describing the receiver coil. Corresponds to DICOM Tag 0018, 1250 `Receive Coil Name`, although not all vendors populate that DICOM Tag, in which case this field can be derived from an appropriate private DICOM field                                      |
+| ReceiveCoilActiveElements     | RECOMMENDED           |               | Information describing the active/selected elements of the receiver coil. This doesn't correspond to a tag in the DICOM ontology. The vendor-defined terminology for active coil elements can go in this field. See an example below the table.                           |
```
what datatype is this?

effigies replied:
All strings in OpenNeuro. I concur.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| MagneticFieldStrength         | RECOMMENDED           |               | Nominal field strength of MR magnet in Tesla. Corresponds to DICOM Tag 0018,0087 `Magnetic Field Strength`                                                                                                                                                                |
+| ReceiveCoilName               | RECOMMENDED           |               | Information describing the receiver coil. Corresponds to DICOM Tag 0018, 1250 `Receive Coil Name`, although not all vendors populate that DICOM Tag, in which case this field can be derived from an appropriate private DICOM field                                      |
+| ReceiveCoilActiveElements     | RECOMMENDED           |               | Information describing the active/selected elements of the receiver coil. This doesn't correspond to a tag in the DICOM ontology. The vendor-defined terminology for active coil elements can go in this field. See an example below the table.                           |
+| GradientSetType               | RECOMMENDED           |               | It should be possible to infer the gradient coil from the scanner model. If not, e.g. because of a custom upgrade or use of a gradient insert set, then the specifications of the actual gradient coil should be reported independently                                   |
```
what datatype is this?

effigies replied:
This is unused, except strings containing a space (`" "`). It reads like a string, though.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| ReceiveCoilName               | RECOMMENDED           |               | Information describing the receiver coil. Corresponds to DICOM Tag 0018, 1250 `Receive Coil Name`, although not all vendors populate that DICOM Tag, in which case this field can be derived from an appropriate private DICOM field                                      |
+| ReceiveCoilActiveElements     | RECOMMENDED           |               | Information describing the active/selected elements of the receiver coil. This doesn't correspond to a tag in the DICOM ontology. The vendor-defined terminology for active coil elements can go in this field. See an example below the table.                           |
+| GradientSetType               | RECOMMENDED           |               | It should be possible to infer the gradient coil from the scanner model. If not, e.g. because of a custom upgrade or use of a gradient insert set, then the specifications of the actual gradient coil should be reported independently                                   |
+| MRTransmitCoilSequence        | RECOMMENDED           |               | This is a relevant field if a non-standard transmit coil is used. Corresponds to DICOM Tag 0018, 9049 `MR Transmit Coil Sequence`                                                                                                                                         |
```
what datatype is this?

effigies replied:
This is unused, except one case of a string containing a space (`" "`). It reads like a string, though.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| ReceiveCoilActiveElements     | RECOMMENDED           |               | Information describing the active/selected elements of the receiver coil. This doesn't correspond to a tag in the DICOM ontology. The vendor-defined terminology for active coil elements can go in this field. See an example below the table.                           |
+| GradientSetType               | RECOMMENDED           |               | It should be possible to infer the gradient coil from the scanner model. If not, e.g. because of a custom upgrade or use of a gradient insert set, then the specifications of the actual gradient coil should be reported independently                                   |
+| MRTransmitCoilSequence        | RECOMMENDED           |               | This is a relevant field if a non-standard transmit coil is used. Corresponds to DICOM Tag 0018, 9049 `MR Transmit Coil Sequence`                                                                                                                                         |
+| MatrixCoilMode                | RECOMMENDED           |               | (If used) A method for reducing the number of independent channels by combining in analog the signals from multiple coil elements. There are typically different default modes when using un-accelerated or accelerated (e.g. GRAPPA, SENSE) imaging                      |
```
what datatype is this?

effigies replied:
This is unused, except strings containing a space (`" "`). It reads like a string, though.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| **Key name**                | **Requirement level** | **Data type** | **Description**                                                                                                                                                                                                                                                   |
+| --------------------------- | --------------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| PulseSequenceType           | RECOMMENDED           | [string][]    | A general description of the pulse sequence used for the scan (i.e. MPRAGE, Gradient Echo EPI, Spin Echo EPI, Multiband gradient echo EPI).                                                                                                                       |
+| ScanningSequence            | RECOMMENDED           |               | Description of the type of data acquired. Corresponds to DICOM Tag 0018, 0020 `Scanning Sequence`.                                                                                                                                                                |
```
what datatype is this?

effigies replied:
In practice, string or array of strings. Both are quite prevalent.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| --------------------------- | --------------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| PulseSequenceType           | RECOMMENDED           | [string][]    | A general description of the pulse sequence used for the scan (i.e. MPRAGE, Gradient Echo EPI, Spin Echo EPI, Multiband gradient echo EPI).                                                                                                                       |
+| ScanningSequence            | RECOMMENDED           |               | Description of the type of data acquired. Corresponds to DICOM Tag 0018, 0020 `Scanning Sequence`.                                                                                                                                                                |
+| SequenceVariant             | RECOMMENDED           |               | Variant of the ScanningSequence. Corresponds to DICOM Tag 0018, 0021 `Sequence Variant`.                                                                                                                                                                          |
```
what datatype is this?

effigies replied:
In practice, string or array of strings. Both are quite prevalent.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| PulseSequenceType           | RECOMMENDED           | [string][]    | A general description of the pulse sequence used for the scan (i.e. MPRAGE, Gradient Echo EPI, Spin Echo EPI, Multiband gradient echo EPI).                                                                                                                       |
+| ScanningSequence            | RECOMMENDED           |               | Description of the type of data acquired. Corresponds to DICOM Tag 0018, 0020 `Scanning Sequence`.                                                                                                                                                                |
+| SequenceVariant             | RECOMMENDED           |               | Variant of the ScanningSequence. Corresponds to DICOM Tag 0018, 0021 `Sequence Variant`.                                                                                                                                                                          |
+| ScanOptions                 | RECOMMENDED           |               | Parameters of ScanningSequence. Corresponds to DICOM Tag 0018, 0022 `Scan Options`.                                                                                                                                                                               |
```
what datatype is this?

effigies replied:
In practice, string or array of strings. Both are quite prevalent.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| ScanningSequence            | RECOMMENDED           |               | Description of the type of data acquired. Corresponds to DICOM Tag 0018, 0020 `Scanning Sequence`.                                                                                                                                                                |
+| SequenceVariant             | RECOMMENDED           |               | Variant of the ScanningSequence. Corresponds to DICOM Tag 0018, 0021 `Sequence Variant`.                                                                                                                                                                          |
+| ScanOptions                 | RECOMMENDED           |               | Parameters of ScanningSequence. Corresponds to DICOM Tag 0018, 0022 `Scan Options`.                                                                                                                                                                               |
+| SequenceName                | RECOMMENDED           |               | Manufacturer's designation of the sequence name. Corresponds to DICOM Tag 0018, 0024 `Sequence Name`.                                                                                                                                                             |
```
what datatype is this?

effigies replied:
All strings in OpenNeuro. I concur.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| ------------------------------ | --------------------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ ...
+| NumberShots                    | RECOMMENDED           | [number][]    | The number of RF excitations need to reconstruct a slice or volume. Please mind that this is not the same as Echo Train Length which denotes the number of lines of k-space collected after an excitation.                                                                                                                           ...
+| ParallelReductionFactorInPlane | RECOMMENDED           | [number][]    | The parallel imaging (e.g, GRAPPA) factor. Use the denominator of the fraction of k-space encoded for each slice. For example, 2 means half of k-space is encoded. Corresponds to DICOM Tag 0018, 9069 `Parallel Reduction Factor In-plane`.                                                                                         ...
+| ParallelAcquisitionTechnique   | RECOMMENDED           |               | The type of parallel imaging used (e.g. GRAPPA, SENSE). Corresponds to DICOM Tag 0018, 9078 `Parallel Acquisition Technique`.                                                                                                                                                                                                        ...
```
what datatype is this?

effigies replied:
String.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| NumberShots                    | RECOMMENDED           | [number][]    | The number of RF excitations need to reconstruct a slice or volume. Please mind that this is not the same as Echo Train Length which denotes the number of lines of k-space collected after an excitation.                                                                                                                           ...
+| ParallelReductionFactorInPlane | RECOMMENDED           | [number][]    | The parallel imaging (e.g, GRAPPA) factor. Use the denominator of the fraction of k-space encoded for each slice. For example, 2 means half of k-space is encoded. Corresponds to DICOM Tag 0018, 9069 `Parallel Reduction Factor In-plane`.                                                                                         ...
+| ParallelAcquisitionTechnique   | RECOMMENDED           |               | The type of parallel imaging used (e.g. GRAPPA, SENSE). Corresponds to DICOM Tag 0018, 9078 `Parallel Acquisition Technique`.                                                                                                                                                                                                        ...
+| PartialFourier                 | RECOMMENDED           |               | The fraction of partial Fourier information collected. Corresponds to DICOM Tag 0018, 9081 `Partial Fourier`.                                                                                                                                                                                                                        ...
```
what datatype is this?

effigies replied:
Number.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| ParallelReductionFactorInPlane | RECOMMENDED           | [number][]    | The parallel imaging (e.g, GRAPPA) factor. Use the denominator of the fraction of k-space encoded for each slice. For example, 2 means half of k-space is encoded. Corresponds to DICOM Tag 0018, 9069 `Parallel Reduction Factor In-plane`.                                                                                         ...
+| ParallelAcquisitionTechnique   | RECOMMENDED           |               | The type of parallel imaging used (e.g. GRAPPA, SENSE). Corresponds to DICOM Tag 0018, 9078 `Parallel Acquisition Technique`.                                                                                                                                                                                                        ...
+| PartialFourier                 | RECOMMENDED           |               | The fraction of partial Fourier information collected. Corresponds to DICOM Tag 0018, 9081 `Partial Fourier`.                                                                                                                                                                                                                        ...
+| PartialFourierDirection        | RECOMMENDED           |               | The direction where only partial Fourier information was collected. Corresponds to DICOM Tag 0018, 9036 `Partial Fourier Direction`.                                                                                                                                                                                                 ...
```
what datatype is this?

effigies replied:
This is unused, except strings containing a space (`" "`). It reads like a string, though.

sappelhoff commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -64,43 +76,43 @@ first "effective" echo and the center of the last "effective" echo.
+| **Key name**                | **Requirement level** | **Data type** | **Description**                                                                                                                                                                                                                                                                                                       |
+| --------------------------- | --------------------- | ------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| FlipAngle                   | RECOMMENDED           | [number][]    | Flip angle for the acquisition, specified in degrees. Corresponds to: DICOM Tag 0018, 1314 `Flip Angle`.                                                                                                                                                                                                              |
+| MultibandAccelerationFactor | RECOMMENDED           |               | The multiband factor, for multiband acquisitions.                                                                                                                                                                                                                                                                     |
```
what datatype is this?

effigies replied:
Number.

sappelhoff reviewed (commented):
I marked those fields that are missing a datatype, please make suggestions.

effigies commented on src/05-derivatives/02-common-data-types.md:
```diff
@@ -71,9 +71,9 @@ The `space` entity may take any value in [Image-Based Coordinate Systems][coords
-| **Key name**     | **Description**                                                                                                                                                                                                                                                                                                          |
-| ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
-| SpatialReference | RECOMMENDED if the derivative is aligned to a standard template listed in [Standard template identifiers][templates]. REQUIRED otherwise. For images with a single reference, the value MUST be a single string. For images with multiple references, such as surface and volume references, a JSON object MUST be used. |
+| **Key name**     | **Requirement level**                                                                                                                     | **Data type**            | **Description**                                                                                                                                                                                    |                ...
```
[Rendering problem](https://bids-specification--605.org.readthedocs.build/en/605/05-derivatives/02-common-data-types.html#spatial-references) with this one. I think you have too many columns in this header line.

```suggestion
| **Key name**     | **Requirement level**                                                                                                                     | **Data type**            | **Description**                                                                                                                                                                                    |
```

sappelhoff replied:
thanks for the catch, updated in code

effigies commented:
Reviewing, but as an initial comment, the **Requirement Level** column takes up a ton of space for 2 bits of information. What if we added abbreviations to use? Some options and evaluations:

| REQUIRED | RECOMMENDED | OPTIONAL | Thoughts |
|--|--|--|--|
| REQ | REC | OPT | Straightforward, REQ and REC look and are pronounced similar |
| RQRD | RCMD | OPTL | Less confusable. Ugly. | 
| Req | Rec | Opt | No longer CAPS, but they are used in isolation, q looks unlike c |
| MUST | SHLD | MAY | Shorter, look like official alternatives, semantically weird |
| MUST | SHOULD | MAY | Same as above, SHOULD looks less weird though |

If we do go with an abbreviation, I would specify it in the common principles, and I would change the header from **Requirement Level** to simply **Level** (also to be described in common principles).

effigies commented on src/05-derivatives/03-imaging.md:
```diff
@@ -131,13 +131,13 @@ and the `Atlas` metadata SHOULD be defined.
+| ------------ | -------------------------------------------------- | -------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
+| RawSources   | REQUIRED                                           | [array][] of [strings][]                           | Same as defined in [Introduction][intro], but elevated from OPTIONAL to REQUIRED.                                                 |
+| Type         | RECOMMENDED                                        | [string][]                                         | Short identifier of the mask. Reserved values: `Brain` - brain mask, `Lesion` - lesion mask, `Face` - face mask, `ROI` - ROI mask |
+| Atlas        | OPTIONAL, RECOMMENDED if `label` entity is defined | [string][]                                         | Which atlas (if any) was used to generate the mask.                                                                               |
```
Seems like unnecessary opportunity for grammatical confusion for no benefit.

```suggestion
| Atlas        | RECOMMENDED if `label` entity is defined | [string][]                                         | Which atlas (if any) was used to generate the mask.                                                                               |
```

effigies commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
 ## Common metadata fields
 
-MR Data described in sections 8.3.x share the following RECOMMENDED metadata
+MR Data described in the [Magnetic Resonance Imaging](./01-magnetic-resonance-imaging-data.md#magnetic-resonance-imaging)
+section (this section) share the following RECOMMENDED metadata
```
I don't see the value in self-linking.

```suggestion
MR Data described in the following sections share the following RECOMMENDED metadata
```

effigies commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -2,56 +2,68 @@
+| EffectiveEchoSpacing           | RECOMMENDED           | [number][]    | The "effective" sampling interval, specified in seconds, between lines in the phase-encoding direction, defined based on the size of the reconstructed image in the phase direction. It is frequently, but incorrectly, referred to as "dwell time" (see `DwellTime` parameter below for actual dwell time). It is required for unwa ...
+| TotalReadoutTime               | RECOMMENDED           | [number][]    | This is actually the "effective" total readout time , defined as the readout duration, specified in seconds, that would have generated data with the given level of distortion. It is NOT the actual, physical duration of the readout train. If `EffectiveEchoSpacing` has been properly computed, it is just `EffectiveEchoSpacing ...
+
+<sup>2</sup>Conveniently, for Siemens' data, this value is easily obtained as
```
I disagree with the possessive here.

```suggestion
<sup>2</sup>Conveniently, for Siemens data, this value is easily obtained as
```

effigies commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -64,43 +76,43 @@ first "effective" echo and the center of the last "effective" echo.
+| ---------------------- | ---------------------------------------------------------------------------------------------------------------- | ------------------------ | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| EchoTime               | RECOMMENDED, but REQUIRED if corresponding fieldmap data is present or the data comes from a multi echo sequence | [number][]               | The echo time (TE) for the acquisition, specified in seconds. Corresponds to DICOM Tag 0018, 0081 `Echo Time` (please note that the DICOM term is in milliseconds not seconds).                                                        ...
+| InversionTime          | RECOMMENDED                                                                                                      | [number][]               | The inversion time (TI) for the acquisition, specified in seconds. Inversion time is the time after the middle of inverting RF pulse to middle of excitation pulse to detect the amount of longitudinal magnetization. Corresponds to  ...
+| SliceTiming            | RECOMMENDED, but REQUIRED for sparse sequences that do not have the `DelayTime` field set                        | [array][] of [numbers][] | The time at which each slice was acquired within each volume (frame) of the acquisition. Slice timing is not slice order -- rather, it is a list of times containing the time (in seconds) of each slice acquisition in relation to th ...
```
i think it might make the most sense to keep the requirement level brief, e.g. `RECOMMENDED/REQUIRED`, and add the `REQUIRED for sparse sequences that do not have the DelayTime field set` to the end of the description.

effigies commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -64,43 +76,43 @@ first "effective" echo and the center of the last "effective" echo.
-| MultibandAccelerationFactor | RECOMMENDED. The multiband factor, for multiband acquisitions. |
+| **Key name**                | **Requirement level** | **Data type** | **Description**                                   |
+| --------------------------- | --------------------- | ------------- | ------------------------------------------------- |
+| MultibandAccelerationFactor | RECOMMENDED           |               | The multiband factor, for multiband acquisitions. |
```
Duplicate of previous table. Probably want to remove it from the previous and keep it here, going by the headers.

effigies commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -275,11 +287,11 @@ JSON file.
+| -------------- | --------------------- | ------------------------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| RepetitionTime | REQUIRED              | [number][]               | The time in seconds between the beginning of an acquisition of one volume and the beginning of acquisition of the volume following it (TR). Please note that this definition includes time between scans (when no data has been acquired) in case of sparse acquisition schemes. This value needs to be consistent with the `pixdim[4]` f ...
+| VolumeTiming   | REQUIRED              | [array][] of [numbers][] | The time at which each volume was acquired during the acquisition. It is described using a list of times referring to the onset of each volume in the BOLD series. The list must have the same length as the BOLD series, and the values must be non-negative and monotonically increasing. This field is mutually exclusive with `Repeti ...
+| TaskName       | REQUIRED              | [string][]               | Name of the task. No two tasks should have the same name. The task label included in the file name is derived from this TaskName field by removing all non-alphanumeric (`[a-zA-Z0-9]`) characters. For example TaskName `faces n-back` will correspond to task label `facesnback`. A RECOMMENDED convention is to name resting state tas ...
```
This should be synced with EEG/MEG/iEEG. I think this is clearer text than those, so I would copy from here.

sappelhoff replied:
Agreed. The iEEG spec had an additional sentence:

> Note this does not have to be a "behavioral task" that subjects perform, but can reflect some information about the conditions present when the data was acquired (e.g., "rest", "sleep", or "seizure").

I pulled this out of the table and put it right below as a note.

sappelhoff commented on src/04-modality-specific-files/08-genetic-descriptor.md:
```diff
@@ -81,14 +82,14 @@ Datasets containing the `Genetics` field in `dataset_description.json` or the
+| **Key name**       | **Requirement level** | **Data type**                          | **Description**                                                                                                                                                                                                                              |
+| ------------------ | --------------------- | -------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| GeneticLevel       | REQUIRED              | [string][] or [array][] of [strings][] | Describes the level of analysis. Values MUST be one of `Genetic`, `Genomic`, `Epigenomic`, `Transcriptomic`, `Metabolomic`, or `Proteomic`.                                                                                                  |
+| AnalyticalApproach | OPTIONAL              | [string][] or [array][] of [strings][] | Methodology or methodologies used to analyse the GeneticLevel. Values MUST be taken from the [database of Genotypes and Phenotypes (dbGaP)][gapsolr] under /Study/Molecular Data Type (e.g., `SNP Genotypes (Array)` or `Methylation (CpG)`. |
```
fair enough, but if okay with you I'd put that off to another PR, because the spec is full of `e.g.` and changing everything would blow up the PR --> those E.G.s I touched *here*, were touched for adding a trailing comma to be consistent.

effigies replied:
Sure.

sappelhoff commented:
Thanks @effigies, I addressed all your comments except the following:

 # shortening the "Requirement Level" column

> Reviewing, but as an initial comment, the Requirement Level column takes up a ton of space for 2 bits of information. What if we added abbreviations to use?

[link to review comment](https://github.com/bids-standard/bids-specification/pull/605#issuecomment-693431234)

I see your point, but I find the new layout aesthetically pleasing, and have not too big of a motivation to decrease the width of the Requirement column manually.

![image](https://user-images.githubusercontent.com/9084751/93441994-312adc80-f8d0-11ea-876b-4597a057662b.png)

Perhaps this could be something to return to once we start generating these tables automatically?

Need some more opinions @tsalo @yarikoptic 

# Keeping the "Requirement level" column brief

> i think it might make the most sense to keep the requirement level brief, e.g. RECOMMENDED/REQUIRED, and add the REQUIRED for sparse sequences that do not have the DelayTime field set to the end of the description.

[link to review comment](https://github.com/bids-standard/bids-specification/pull/605#discussion_r489497179)

Personally, I find it nicer to read everything about requirements in the requirement level column. It makes more sense to me logically, but you do have a point that a very wide "requirement level" column looks ugly ... especially when there are few rows that need the width, and most rows could actually get along with having a very short width.

**Yet** in rendered form, this looks quite okay and nice IMHO:

![image](https://user-images.githubusercontent.com/9084751/93443001-75b67800-f8d0-11ea-8938-1221403bca34.png)

So I'd be in favor of keeping it as it is

effigies commented on src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md:
```diff
@@ -596,3 +606,16 @@ file to obtain scanning parameters. \_epi files can be a 3D or 4D - in the
+
+<!-- Link Definitions -->
+
+[deprecated]: https://en.wikipedia.org/wiki/Deprecation
```
For a separate PR: Might be good to have a note on deprecation in the spec itself. Something like:

> DEPRECATED - A "deprecated" entity or metadata field SHOULD NOT be used in the generation of new datasets. It remains in the standard in order to preserve the interpretability of existing datasets. Validating software SHOULD warn when deprecated practices are detected and provide a suggestion for updating the dataset to preserve the curator's intent.

effigies commented on src/04-modality-specific-files/04-intracranial-electroencephalography.md:
```diff
@@ -338,30 +343,30 @@ listed below.
-| manufacturer | OPTIONAL. Recommended field to specify the manufacturer for each electrode. Can be used if electrodes were manufactured by more than one company.                    |
-| group        | OPTIONAL. Optional field to specify the group that the electrode is a part of. Note that any group specified here should match a group specified in `_channels.tsv`. |
-| hemisphere   | OPTIONAL. Optional field to specify the hemisphere in which the electrode is placed, one of `['L' or 'R']` (use capital).                                            |
+| **Column name**  | **Requirement level** | **Description**                                                                                                                                               |
+| ---------------- | --------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| material         | RECOMMENDED           | Material of the electrodes.                                                                                                                                   |
+| manufacturer     | RECOMMENDED           | Recommended field to specify the manufacturer for each electrode. Can be used if electrodes were manufactured by more than one company.                       |
+| group            | RECOMMENDED           | RECOMMENDED field to specify the group that the electrode is a part of. Note that any group specified here should match a group specified in `_channels.tsv`. |
+| hemisphere       | RECOMMENDED           | RECOMMENDED field to specify the hemisphere in which the electrode is placed, one of `['L' or 'R']` (use capital).                                            |
```
Tried to remove the right number of spaces/dashes, but I might have gotten it wrong.

```suggestion
| **Column name**  | **Requirement level** | **Description**                                                                                                                  |
| ---------------- | --------------------- | --------------------------------------------------------------------------------------------------------------------------------- |
| material         | RECOMMENDED           | Material of the electrodes.                                                                                                      |
| manufacturer     | RECOMMENDED           | The manufacturer for each electrode. Can be used if electrodes were manufactured by more than one company.                       |
| group            | RECOMMENDED           | The group that the electrode is a part of. Note that any group specified here should match a group specified in `_channels.tsv`. |
| hemisphere       | RECOMMENDED           | The hemisphere in which the electrode is placed, one of `['L' or 'R']` (use capital).                                            |
```

effigies reviewed (approved):
Looks good. One small suggestion.

effigies commented:
Oh, and FWIW I agree that the tables look fine in HTML. In the PDF, the description is pushed quite far to the right.

sappelhoff commented:
@tsalo could you please have a final look again? I'd like to get this merged as soon as possible to avoid merge conflicts ... this PR is touching almost every table in the spec after all :-)

tsalo commented on src/04-modality-specific-files/02-magnetoencephalography.md:
```diff
@@ -287,47 +287,47 @@ EEG, head localization coils, and anatomical landmarks.
-| IntendedFor | OPTIONAL. Path or list of path relative to the subject subfolder pointing to the structural MRI, possibly of different types if a list is specified, to be used with the MEG recording. The path(s) need(s) to use forward slashes instead of backward slashes (e.g. `ses-/anat/sub-01_T1w.nii.gz`). |
+| **Key name** | **Requirement level** | **Data type**                          | **Description**                                                                                                                                                                                                                                                                             |
+| ------------ | --------------------- | -------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| IntendedFor  | OPTIONAL              | [string][] or [array][] of [strings][] | Path or list of path relative to the subject subfolder pointing to the structural MRI, possibly of different types if a list is specified, to be used with the MEG recording. The path(s) need(s) to use forward slashes instead of backward slashes (e.g., `ses-/anat/sub-01_T1w.nii.gz`). |
```
It looks like the `strings` link isn't defined in this page.

sappelhoff replied:
ahh good catch, thanks! Fixed in https://github.com/bids-standard/bids-specification/pull/605/commits/e87497cb9fe7691d61518b1c0efcc193f396fc40

tsalo reviewed (changes requested):
I noticed one minor issue but other than that everything looks great to me.

sappelhoff commented:
Thanks for the reviews, this is going in now.
